### PR TITLE
feat: Extend currency definitions & add definitions query

### LINKED ISSUE #228: [Oracle] Provide info for the currencies pertaining to the protocol
The use case to enable is the following: The App obtains the protocols from the Admin, incl. the oracles. Then query them to get the info subject of this issue.

We need the (ticker, symbol@Nolus, symbol@DexNetwork, decimals/exponent) of each of the Lpn, lease currencies, and extra payment currencies.

Payment currencies = union of the Lpn, lease currencies, and the extra payment currencies (if any, for example, Nls).

The ticker must be the same as the one in `currencies.json` at the Dex network. The only exception, cause it needs a migration, is USDC at Osmosis.

### PR DESCRIPTION
**Pull request by KirilMihaylov**
Note: Implementation only lists currently configured currencies rather than all software supported.

Closes #228

### CODE REVIEW

File: platform/packages/currency/src/lib.rs [hunk 1]
```diff
@@ -40,6 +40,12 @@ pub trait Currency: Copy + Ord + Default + Debug + 'static {
 
     /// Symbol at the Dex network
     const DEX_SYMBOL: SymbolStatic;
+
+    /// Exponent on which the whole unit was raised to get the currency's base
+    /// unit represented by the trait.
+    ///
+    /// Example: `(10 * DECIMAL_PLACES) uUSDC = 1 USDC`
+    const DECIMAL_PLACES: u8;
```
Comment (gmanev7) [inline platform/packages/currency/src/lib.rs:48]: [everywhere] "places" is not so suitable, please rename to "digits"

File: protocol/contracts/oracle/src/api/contract.rs [hunk 2]
```diff
@@ -113,6 +115,17 @@ pub struct DispatchAlarmsResponse(pub AlarmsCount);
 
 pub type SupportedCurrencyPairsResponse = Vec<SwapLeg>;
 
+pub type CurrencyDefinitionsResponse = Vec<CurrencyDefinition>;
+
+#[derive(Serialize, Deserialize)]
+#[serde(deny_unknown_fields, rename_all = "snake_case")]
+pub struct CurrencyDefinition {
```
Comment (gmanev7) [inline protocol/contracts/oracle/src/api/contract.rs:122]: -> Currency

File: protocol/contracts/oracle/src/api/contract.rs [hunk 3]
```diff
@@ -76,8 +76,10 @@ pub enum QueryMsg {
     Price {
         currency: SymbolOwned,
     },
-    // returns a list of supported denom pairs
+    /// Lists configured swap pairs
     SupportedCurrencyPairs {},
+    /// Lists configured currency definitions
+    CurrencyDefinitions {},
```
Comment (gmanev7) [inline protocol/contracts/oracle/src/api/contract.rs:82]: -> Currencies

File: protocol/contracts/oracle/src/state/supported_pairs.rs [hunk 4]
```diff
@@ -178,6 +189,26 @@ where
...
+struct DefinitionVisitor;
+
+impl AnyVisitor for DefinitionVisitor {
+    type Output = CurrencyDefinition;
+
+    type Error = platform::never::Never;
+
+    fn on<C>(self) -> AnyVisitorResult<Self>
+    where
+        C: Currency + Serialize + DeserializeOwned,
```
Comment (gmanev7) [inline protocol/contracts/oracle/src/state/supported_pairs.rs:201]: Are you sure we need `Serialize + Deserialize`

File: protocol/contracts/oracle/src/api/contract.rs [hunk 5]
```diff
@@ -113,6 +115,17 @@ pub struct DispatchAlarmsResponse(pub AlarmsCount);
...
+pub type CurrencyDefinitionsResponse = Vec<CurrencyDefinition>;
+
+#[derive(Serialize, Deserialize)]
+#[serde(deny_unknown_fields, rename_all = "snake_case")]
+pub struct CurrencyDefinition {
+    pub ticker: String,
+    pub bank_symbol: String,
+    pub dex_symbol: String,
+    pub decimal_places: u8,
+}
```
Comment (gmanev7) [inline protocol/contracts/oracle/src/api/contract.rs:127]: we also need to know the group it directly is a member of. They are Native, Lpns, Lease, and PaymentOnlyGroup

File: protocol/packages/currencies/src/lease/osmosis.rs [hunk 6]
```diff
@@ -274,7 +274,7 @@ define_symbol! {
         },
     }
 }
-define_currency!(StkAtom, STK_ATOM);
+define_currency!(StkAtom, STK_ATOM, 8);
```
Comment (gmanev7) [inline protocol/packages/currencies/src/lease/osmosis.rs:277]: please double-check it is eight! The definition in currencies.json is 6

File: protocol/contracts/oracle/src/state/supported_pairs.rs (hunk 4 above)
Comment (KirilMihaylov) [reply protocol/contracts/oracle/src/state/supported_pairs.rs:201]: It came with the signature as far as I am aware. Most likely the IDE generated the method for me, thus it's already a bound on the trait. I can remove it here, but it won't make a big difference overall.

File: platform/packages/currency/src/lib.rs [hunk 7]
```diff
@@ -40,6 +40,12 @@ pub trait Currency: Copy + Ord + Default + Debug + 'static {
 
     /// Symbol at the Dex network
     const DEX_SYMBOL: SymbolStatic;
+
+    /// Exponent on which the whole unit was raised to get the currency's base
+    /// unit represented by the trait.
+    ///
+    /// Example: `(10 * DECIMAL_DIGITS) uUSDC = 1 USDC`
```
Comment (gmanev7) [inline platform/packages/currency/src/lib.rs:47]: * -> ^

File: protocol/contracts/oracle/src/api/contract.rs [hunk 8]
```diff
@@ -76,8 +76,10 @@ pub enum QueryMsg {
     Price {
         currency: SymbolOwned,
     },
-    // returns a list of supported denom pairs
+    /// Lists configured swap pairs
     SupportedCurrencyPairs {},
+    /// Lists configured currency definitions
```
Comment (gmanev7) [inline protocol/contracts/oracle/src/api/contract.rs:81]: currency definitions -> currencies

File: protocol/contracts/oracle/src/state/supported_pairs.rs [hunk 9]
```diff
@@ -160,6 +160,38 @@ where
...
+                        )
+                    })
+                    .or_else(|_| {
+                        Tickers.maybe_visit_any::<currencies::LeaseGroup, _>(
+                            &node.value().target,
+                            CurrencyVisitor(api::CurrencyGroup::Lease),
+                        )
+                    })
+                    .or_else(|_| {
+                        Tickers.maybe_visit_any::<currencies::PaymentGroup, _>(
```
Comment (gmanev7) [inline protocol/contracts/oracle/src/state/supported_pairs.rs:185]: currencies::PaymentOnly

File: protocol/contracts/oracle/src/state/supported_pairs.rs (hunk 4 above)
Comment (gmanev7) [reply protocol/contracts/oracle/src/state/supported_pairs.rs:222]: how big the difference is we cannot measure but for sure we have to limit the required bounds to their minimum

File: protocol/contracts/oracle/src/state/supported_pairs.rs [hunk 10]
```diff
@@ -160,6 +160,38 @@ where
             })
     }
 
+    pub fn currency_definitions(&self) -> impl Iterator<Item = api::Currency> + '_ {
```
Comment (gmanev7) [inline protocol/contracts/oracle/src/state/supported_pairs.rs:163]: this function deserves at least a few unit tests

File: protocol/contracts/oracle/src/state/supported_pairs.rs (hunk 4 above)
Comment (KirilMihaylov) [reply protocol/contracts/oracle/src/state/supported_pairs.rs:222]: I'll remove it from here, but it's the trait that needs to be reworked is what I meant. ;)

File: protocol/contracts/oracle/src/state/supported_pairs.rs [hunk 11]
```diff
@@ -178,11 +211,32 @@ where
...
+    where
+        C: Currency,
+    {
+        Ok(api::Currency {
+            ticker: C::TICKER.into(),
+            bank_symbol: C::BANK_SYMBOL.into(),
+            dex_symbol: C::DEX_SYMBOL.into(),
+            decimal_digits: C::DECIMAL_DIGITS,
+            group: self.0,
+        })
```
Comment (gmanev7) [inline protocol/contracts/oracle/src/state/supported_pairs.rs:231]: define this as a fn of api::Currency and reuse in the tests below

**Review by gmanev7 — APPROVED** [review]

**Pull request merged**